File: core/src/com/sheepsgohome/screens/SettingsPlayerScreen.java
package com.sheepsgohome.screens;

import com.badlogic.gdx.Game;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Screen;
import com.badlogic.gdx.graphics.GL20;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.badlogic.gdx.scenes.scene2d.Stage;
import com.badlogic.gdx.scenes.scene2d.ui.Image;
import com.badlogic.gdx.scenes.scene2d.ui.Label;
import com.badlogic.gdx.scenes.scene2d.ui.Table;
import com.badlogic.gdx.scenes.scene2d.ui.TextButton;
import com.badlogic.gdx.scenes.scene2d.ui.TextField;
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener;
import com.badlogic.gdx.utils.viewport.StretchViewport;
import com.sheepsgohome.GameData;
import com.sheepsgohome.leaderboard.LeaderBoardCallback;
import com.sheepsgohome.dialogs.MessageDialog;
import com.sheepsgohome.dialogs.OkDialog;
import com.sheepsgohome.leaderboard.LeaderBoardResult;

import static com.badlogic.gdx.graphics.Texture.TextureFilter;
import static com.sheepsgohome.GameData.CAMERA_HEIGHT;
import static com.sheepsgohome.GameData.CAMERA_WIDTH;
import static com.sheepsgohome.GameData.Loc;
import static com.sheepsgohome.GameSkins.skin;

/**
 * Created by LittleLight on 11.1.2015.
 */
public class SettingsPlayerScreen implements Screen, LeaderBoardCallback {

    private static final float BUTTON_WIDTH = 100;
    private static final float BUTTON_SMALL_WIDTH = 50;
    private Stage stage;
    private Table table;

    private TextButton buttonRegister;
    private TextButton buttonSave;
    private TextButton buttonBack;

    private TextField playerName;

    private Label title;
    private Label playerNameTitle;

    private MessageDialog messageDialog = null;

    Texture texture;
    Image bgImage;

    @Override
    public void show() {
        buttonRegister = new TextButton(Loc.get("register"), skin);
        buttonSave = new TextButton(Loc.get("save"), skin);
        buttonBack = new TextButton(Loc.get("back"), skin);
        title = new Label(Loc.get("player"), skin, "menuTitle");
        playerNameTitle = new Label(Loc.get("player.name"), skin, "default");

        playerName = new TextField("", skin);
        playerName.setMaxLength(16);
        playerName.setAlignment(1);
        playerName.setText(GameData.PLAYER_NAME);

        texture = new Texture("menu_background.png");
        bgImage = new Image(texture);

        float multiplier = 2;
        stage = new Stage(new StretchViewport(CAMERA_WIDTH * multiplier, CAMERA_HEIGHT * multiplier));

        table = new Table();
//        table.setDebug(true);

        //click listeners
        buttonSave.addListener(new ClickListener() {
            @Override
            public void clicked(InputEvent event, float x, float y) {
                GameData.PLAYER_NAME = validatePlayerName();
                GameData.SavePreferences();
                ((Game) Gdx.app.getApplicationListener()).setScreen(new SettingsScreen());
            }
        });

        buttonBack.addListener(new ClickListener() {
            @Override
            public void clicked(InputEvent event, float x, float y) {
                ((Game) Gdx.app.getApplicationListener()).setScreen(new SettingsScreen());
            }
        });


        final SettingsPlayerScreen __this = this;
        buttonRegister.addListener(new ClickListener() {
            @Override
            public void clicked(InputEvent event, float x, float y) {

                String nick = validatePlayerName();

                if (nick.equals("")) {
                    OkDialog dialog = new OkDialog(Loc.get("empty.player.name"), skin, "dialog");
                    dialog.setPrefHeight(60);
                    dialog.show(stage);
                } else {
                    GameData.leaderboard.register(
                            GameData.functions.getDeviceId(),
                            nick,
                            GameData.LEVEL,
                            GameData.functions.getCountryCode(),
                            __this);
                }
            }
        });


        //table
        table.setFillParent(true);

        title.setFontScale(GameData.SETTINGS_TITLE_FONT_SCALE);
        table.add(title).top().colspan(2).row();

        Table contentTable = new Table();
        playerNameTitle.setFontScale(GameData.SETTINGS_ITEM_FONT_SCALE);
        contentTable.add(playerNameTitle).expandX().row();
        contentTable.add(playerName).size(170, BUTTON_SMALL_WIDTH / 2).padTop(5).padBottom(5).row();

        float BUTTON_REGISTER_WIDTH = BUTTON_WIDTH;
        contentTable.add(buttonRegister).size(BUTTON_REGISTER_WIDTH, BUTTON_REGISTER_WIDTH / 2).row();


        table.add(contentTable).expand().colspan(2).top().row();

        table.add(buttonSave).size(BUTTON_SMALL_WIDTH, BUTTON_SMALL_WIDTH / 2).bottom().right();
        table.add(buttonBack).size(BUTTON_SMALL_WIDTH, BUTTON_SMALL_WIDTH / 2).bottom().left().row();

        stage.addActor(bgImage);
        stage.addActor(table);

        Gdx.input.setInputProcessor(stage);

        bgImage.setWidth(CAMERA_WIDTH * multiplier);
        bgImage.setHeight(CAMERA_HEIGHT * multiplier);
        texture.setFilter(TextureFilter.Linear, TextureFilter.Linear);
    }

    private String validatePlayerName() {
        String nick = playerName.getText().replaceAll("[^\\w\\_\\-]", "");
        playerName.setText(nick);
        return nick;
    }

    @Override
    public void render(float delta) {
        Gdx.gl.glClearColor(0.5f, 0.5f, 0.5f, 1);
        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT);
        stage.act();
        stage.draw();
    }

    @Override
    public void resize(int width, int height) {
        stage.getViewport().update(width, height, true);
    }

    @Override
    public void pause() {

    }

    @Override
    public void resume() {

    }

    @Override
    public void hide() {
        dispose();
    }

    @Override
    public void dispose() {
        stage.dispose();
        texture.dispose();
    }

    /**
     * LeaderBoard callback
     */
    private void hideMessageDialog() {
        if (messageDialog != null) {
            messageDialog.hide();
        }
    }

    @Override
    public void connecting() {
        messageDialog = new MessageDialog(Loc.get("connecting"), skin, "dialog");
        messageDialog.setPrefHeight(40);
        messageDialog.show(stage);
    }

    @Override
    public void connectionToDatabaseFailed() {
        hideMessageDialog();

        OkDialog dialog = new OkDialog(Loc.get("connection.to.database.failed"), skin, "dialog");
        dialog.setPrefHeight(60);
        dialog.show(stage);
    }

    @Override
    public void invalidData() {
        hideMessageDialog();

        OkDialog dialog = new OkDialog(Loc.get("invalid.data"), skin, "dialog");
        dialog.setPrefHeight(80);
        dialog.show(stage);
    }

    @Override
    public void nickAlreadyInUse() {
        hideMessageDialog();

        OkDialog dialog = new OkDialog(Loc.get("player.name.already.in.use"), skin, "dialog");
        dialog.setPrefHeight(90);
        dialog.show(stage);
    }


    @Override
    public void success() {
        hideMessageDialog();

        OkDialog dialog = new OkDialog(Loc.get("player.name.registration.success"), skin, "dialog");
        dialog.setPrefHeight(70);
        dialog.show(stage);

        GameData.PLAYER_NAME = playerName.getText();
        GameData.SavePreferences();

    }

    @Override
    public void failure() {
        hideMessageDialog();

        OkDialog dialog = new OkDialog(Loc.get("unknown.failure"), skin, "dialog");
        dialog.setPrefHeight(60);
        dialog.show(stage);
    }

    @Override
    public void failedToInitializeMD5() {
        hideMessageDialog();

        OkDialog dialog = new OkDialog(Loc.get("md5.init.failed"), skin, "dialog");
        dialog.setPrefHeight(95);
        dialog.show(stage);
    }

    @Override
    public void connectionFailed() {
        hideMessageDialog();

        OkDialog dialog = new OkDialog(Loc.get("connection.failed"), skin, "dialog");
        dialog.setPrefHeight(60);
        dialog.show(stage);
    }

    @Override
    public void connectionCanceled() {
        hideMessageDialog();

        OkDialog dialog = new OkDialog(Loc.get("connection.canceled"), skin, "dialog");
        dialog.setPrefHeight(85);
        dialog.show(stage);
    }

    @Override
    public void unregisteredUser() {

    }

    @Override
    public void leaderboardResult(LeaderBoardResult result) {

    }
}

File: core/src/com/sheepsgohome/screens/SettingsPlayerScreen.kt
package com.sheepsgohome.screens

import com.badlogic.gdx.Game
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.Screen
import com.badlogic.gdx.graphics.GL20
import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.graphics.Texture.TextureFilter
import com.badlogic.gdx.scenes.scene2d.InputEvent
import com.badlogic.gdx.scenes.scene2d.Stage
import com.badlogic.gdx.scenes.scene2d.ui.*
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener
import com.badlogic.gdx.utils.viewport.StretchViewport
import com.sheepsgohome.GameSkins.skin
import com.sheepsgohome.dialogs.MessageDialog
import com.sheepsgohome.dialogs.OkDialog
import com.sheepsgohome.leaderboard.LeaderBoard
import com.sheepsgohome.leaderboard.LeaderBoardCallback
import com.sheepsgohome.leaderboard.LeaderBoardResult
import com.sheepsgohome.shared.GameData
import com.sheepsgohome.shared.GameData.CAMERA_HEIGHT
import com.sheepsgohome.shared.GameData.CAMERA_WIDTH
import com.sheepsgohome.shared.GameData.loc

class SettingsPlayerScreen : Screen, LeaderBoardCallback {

    private val leaderBoard = LeaderBoard.instance

    private lateinit var stage: Stage
    private lateinit var table: Table

    private lateinit var buttonRegister: TextButton
    private lateinit var buttonSave: TextButton
    private lateinit var buttonBack: TextButton

    private lateinit var playerName: TextField

    private lateinit var title: Label
    private lateinit var playerNameTitle: Label

    private var messageDialog: MessageDialog? = null

    private lateinit var texture: Texture
    private lateinit var bgImage: Image

    override fun show() {
        buttonRegister = TextButton(loc.get("register"), skin)
        buttonSave = TextButton(loc.get("save"), skin)
        buttonBack = TextButton(loc.get("back"), skin)
        title = Label(loc.get("player"), skin, "menuTitle")
        playerNameTitle = Label(loc.get("player.name"), skin, "default")

        playerName = TextField("", skin)
        playerName.maxLength = 16
        playerName.setAlignment(1)
        playerName.text = GameData.PLAYER_NAME

        texture = Texture("menu_background.png")
        bgImage = Image(texture)

        val multiplier = 2f
        stage = Stage(StretchViewport(CAMERA_WIDTH * multiplier, CAMERA_HEIGHT * multiplier))

        table = Table()

        //click listeners
        buttonSave.addListener(object : ClickListener() {
            override fun clicked(event: InputEvent?, x: Float, y: Float) {
                GameData.PLAYER_NAME = validatePlayerName()
                GameData.savePreferences()
                (Gdx.app.applicationListener as Game).screen = SettingsScreen()
            }
        })

        buttonBack.addListener(object : ClickListener() {
            override fun clicked(event: InputEvent?, x: Float, y: Float) {
                (Gdx.app.applicationListener as Game).screen = SettingsScreen()
            }
        })


        val __this = this
        buttonRegister.addListener(object : ClickListener() {
            override fun clicked(event: InputEvent?, x: Float, y: Float) {

                val nick = validatePlayerName()

                if (nick == "") {
                    val dialog = OkDialog(loc.get("empty.player.name"), skin, "dialog")
                    dialog.prefHeight = 60f
                    dialog.show(stage)
                } else {
                    leaderBoard.register(
                            GameData.functions.deviceId,
                            nick,
                            GameData.LEVEL,
                            GameData.functions.countryCode,
                            __this)
                }
            }
        })


        //table
        table.setFillParent(true)

        title.setFontScale(GameData.SETTINGS_TITLE_FONT_SCALE)
        table.add<Label>(title).top().colspan(2).row()

        val contentTable = Table()
        playerNameTitle.setFontScale(GameData.SETTINGS_ITEM_FONT_SCALE)
        contentTable.add<Label>(playerNameTitle).expandX().row()
        contentTable.add<TextField>(playerName).size(170f, BUTTON_SMALL_WIDTH / 2).padTop(5f).padBottom(5f).row()

        val BUTTON_REGISTER_WIDTH = BUTTON_WIDTH
        contentTable.add<TextButton>(buttonRegister).size(BUTTON_REGISTER_WIDTH, BUTTON_REGISTER_WIDTH / 2).row()


        table.add(contentTable).expand().colspan(2).top().row()

        table.add<TextButton>(buttonSave).size(BUTTON_SMALL_WIDTH, BUTTON_SMALL_WIDTH / 2).bottom().right()
        table.add<TextButton>(buttonBack).size(BUTTON_SMALL_WIDTH, BUTTON_SMALL_WIDTH / 2).bottom().left().row()

        stage.addActor(bgImage)
        stage.addActor(table)

        Gdx.input.inputProcessor = stage

        bgImage.width = CAMERA_WIDTH * multiplier
        bgImage.height = CAMERA_HEIGHT * multiplier
        texture.setFilter(TextureFilter.Linear, TextureFilter.Linear)
    }

    private fun validatePlayerName(): String {
        val nick = playerName.text.replace("[^\\w_\\-]".toRegex(), "")
        playerName.text = nick
        return nick
    }

    override fun render(delta: Float) {
        Gdx.gl.glClearColor(0.5f, 0.5f, 0.5f, 1f)
        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT)
        stage.act()
        stage.draw()
    }

    override fun resize(width: Int, height: Int) {
        stage.viewport.update(width, height, true)
    }

    override fun pause() {

    }

    override fun resume() {

    }

    override fun hide() {
        dispose()
    }

    override fun dispose() {
        stage.dispose()
        texture.dispose()
    }

    /**
     * LeaderBoard callback
     */
    private fun hideMessageDialog() {
        messageDialog?.hide()
    }

    override fun connecting() {
        messageDialog = MessageDialog(loc.get("connecting"), skin, "dialog")
        messageDialog?.prefHeight = 40f
        messageDialog?.show(stage)
    }

    override fun connectionToDatabaseFailed() {
        hideMessageDialog()

        val dialog = OkDialog(loc.get("connection.to.database.failed"), skin, "dialog")
        dialog.prefHeight = 60f
        dialog.show(stage)
    }

    override fun invalidData() {
        hideMessageDialog()

        val dialog = OkDialog(loc.get("invalid.data"), skin, "dialog")
        dialog.prefHeight = 80f
        dialog.show(stage)
    }

    override fun nickAlreadyInUse() {
        hideMessageDialog()

        val dialog = OkDialog(loc.get("player.name.already.in.use"), skin, "dialog")
        dialog.prefHeight = 90f
        dialog.show(stage)
    }


    override fun success() {
        hideMessageDialog()

        val dialog = OkDialog(loc.get("player.name.registration.success"), skin, "dialog")
        dialog.prefHeight = 70f
        dialog.show(stage)

        GameData.PLAYER_NAME = playerName.text
        GameData.savePreferences()

    }

    override fun failure() {
        hideMessageDialog()

        val dialog = OkDialog(loc.get("unknown.failure"), skin, "dialog")
        dialog.prefHeight = 60f
        dialog.show(stage)
    }

    override fun failedToInitializeMD5() {
        hideMessageDialog()

        val dialog = OkDialog(loc.get("md5.init.failed"), skin, "dialog")
        dialog.prefHeight = 95f
        dialog.show(stage)
    }

    override fun connectionFailed() {
        hideMessageDialog()

        val dialog = OkDialog(loc.get("connection.failed"), skin, "dialog")
        dialog.prefHeight = 60f
        dialog.show(stage)
    }

    override fun connectionCanceled() {
        hideMessageDialog()

        val dialog = OkDialog(loc.get("connection.canceled"), skin, "dialog")
        dialog.prefHeight = 85f
        dialog.show(stage)
    }

    override fun unregisteredUser() {

    }

    override fun leaderboardResult(result: LeaderBoardResult) {

    }

    companion object {
        private val BUTTON_WIDTH = 100f
        private val BUTTON_SMALL_WIDTH = 50f
    }
}
